educator/src/Dscp/DB/SQLite/Util.hs

```diff
@@ -1,79 +1,213 @@
+    -> pk
+    -> DBT t w m (Maybe res)
+selectByPk mapper tbl key =
+    fmap fetchOne $
+    runSelect . Beam.select $
+    filter_ (valPk_ key `references_`) $
+    all_ tbl
+  where
+    fetchOne [] = Nothing
+    fetchOne l  = Just (mapper $ oneOrError l)
+
+-- | Quick way to check whether an entiry refered by the given primary key exists.
+existsWithPk
+    :: _
+    => Beam.DatabaseEntity be db (TableEntity table) -> pk -> DBT t w m Bool
+existsWithPk tbl key =
+    checkExists $ do
+        row <- all_ tbl
+        guard_ (Beam.pk row ==. valPk_ key)
+
+-- | Quick way to delete a single entiry refered by the given primary key.
+deleteByPk
+    :: (MonadIO m, _)
+    => Beam.DatabaseEntity be db (TableEntity table)
+    -> pk
+    -> DBT t 'Writing m ()
+deleteByPk tbl key = runDelete $ delete tbl (valPk_ key `references_`)
+
+-- | SQL CURRENT_TIMESTAMP function.
+currentTimestampUtc_
+    :: forall ctxt syntax s.
+       Beam.IsSql92ExpressionSyntax syntax
+    => Beam.QGenExpr ctxt syntax s UTCTime
+currentTimestampUtc_ =
+    -- The behavior should not depend much on the backend.
+    -- CURRENT_TIMESTAMP is nicely fetched both for "datetime" and "timestamp" types.
+    coerce (Beam.currentTimestamp_ @syntax)
+
+-- | Make a filtering predicate checking for values match.
+filterMatches_
+    :: _
+    => Maybe a -> QGenExpr syntax ctx s a -> QGenExpr syntax ctx s Bool
+filterMatches_ = \case
+    Nothing -> \_ -> val_ True
+    Just x  -> (==. val_ x)
+
+-- | Make a filtering predicate checking for values match.
+filterMatchesPk_
+    :: _
+    => Maybe a
+    -> PrimaryKey row (QGenExpr syntax ctx s)
+    -> QGenExpr syntax ctx s Bool
+filterMatchesPk_ = \case
+    Nothing -> \_ -> val_ True
+    Just x  -> (==. valPk_ x)
+
+-- | Make a filtering guard checking for values match.
+guardFilterMatches_
+    :: _
+    => Maybe a
+    -> QGenExpr syntax ctx s a
+    -> Beam.Q select db s ()
+guardFilterMatches_ = \case
+    Nothing -> \_ -> pass
+    Just x  -> guard_ . (==. val_ x)
+
+-- | Make a filtering guard checking for primary keys match.
+guardFilterMatchesPk_
+    :: _
+    => Maybe a
+    -> PrimaryKey row (QGenExpr syntax ctx s)
+    -> Beam.Q select db s ()
+guardFilterMatchesPk_ = \case
+    Nothing -> \_ -> pass
+    Just x  -> guard_ . (==. valPk_ x)
+
+-- | Take the next primary key for the given table, can be used for further insert.
+-- In [DSCP-388] we probably remove this function for the sake of specialized methods
+-- to do the similar thing.
+getNextPrimaryKey
+    :: (PrimaryKeyWrapper (PrimaryKey table Identity) keyInner
+       ,MonadIO m
+       ,Num keyInner
+       ,Beam.Table table
+       ,Beam.Database be db
+       ,Beam.FromBackendRow be keyInner
+       ,be ~ Sqlite
+       ,_)
+    => Beam.DatabaseEntity be db (TableEntity table)
+    -> DBT 'WithinTx 'Writing m keyInner
+getNextPrimaryKey tbl = do
+    res <- runSelect . select $
+        aggregate_ (Beam.max_ . unpackPk . pk_) (all_ tbl)
```

Comment:
Wat, why not do it via `orderBy_` and `limit_`?...

Reply:
Well, AFAIU in the btree index rows are already ordered by the index key (and Sqlite's primary keys use btree indices IIRC), so queries with `order ...` and `limit k` take `O(k)` time, while this query will expand to `MAX(ALL ...)` SQL query (http://hackage.haskell.org/package/beam-core-0.7.2.2/docs/Database-Beam-Query.html#v:max_), and I'm not sure whether or not SQLite optimizes such queries.

Reply:
Ah, okay, turns out it does optimize such queries (https://www.sqlite.org/optoverview.html), so it's legit. If you did know this in advance, then no diss.

Reply:
I didn't, nice to know )
Quick googling shows that Postgres seems to optimize it as well.